src/compiler/ruby_generator.cc

```diff
@@ -37,16 +38,65 @@ using std::vector;
 namespace grpc_ruby_generator {
 namespace {
 
+// returns the full canonical message name
+grpc::string GetCanonicalMessageType(const Descriptor* msg) {
+  const Descriptor* top_level_msg;
+  const FileDescriptor* file_containing_msg;
+  grpc::string msg_full_name;
+  grpc::string msg_name;
+  grpc::string msg_proto_pkg;
+  grpc::string resolved_namespace;
+  grpc::string res(msg->full_name());
+
+  // If msg is nested, find the topmost message
+  top_level_msg = msg;
+  while (top_level_msg->containing_type()) {
+    top_level_msg = top_level_msg->containing_type();
+  }
+  file_containing_msg = top_level_msg->file();
+
+  msg_full_name = top_level_msg->full_name();
+  msg_name = top_level_msg->name();
+  msg_proto_pkg =
+      msg_full_name.substr(0, msg_full_name.length() - msg_name.length());
+
+  if (file_containing_msg->options().has_ruby_package()) {
+    resolved_namespace = file_containing_msg->options().ruby_package();
+  }
```

Comment:
style nit: can we move `else` up inline with the brace. Here and on line 78 as well